File: app/domain/schemas.py
"""Pydantic schemas for request validation and response serialization.

Provides input schemas (``CameraCreate``, ``CameraUpdate``), read schemas
(``CameraRead``, ``SnapshotRead``, ``SnapshotAnalysisRead``), and report/result
schemas used across the API layer including ML analysis.
"""

from datetime import date, datetime
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, StringConstraints


class CameraCreate(BaseModel):
    """Validate payload for creating a new camera."""

    name: Annotated[str, StringConstraints(min_length=1, strip_whitespace=True)]
    host: Annotated[str, StringConstraints(min_length=1, strip_whitespace=True)]
    port: int = Field(default=80, ge=1, le=65535)
    username: str = ""
    password: str = ""
    profile_token: str | None = None
    snapshot_url: str | None = None
    interval_seconds: int = Field(default=60, ge=10)
    enabled: bool = True


class CameraUpdate(BaseModel):
    """Validate payload for partially updating an existing camera."""

    name: Annotated[str, StringConstraints(min_length=1, strip_whitespace=True)] | None = None
    host: Annotated[str, StringConstraints(min_length=1, strip_whitespace=True)] | None = None
    port: int | None = Field(default=None, ge=1, le=65535)
    username: str | None = None
    password: str | None = None
    profile_token: str | None = None
    snapshot_url: str | None = None
    interval_seconds: int | None = Field(default=None, ge=10)
    enabled: bool | None = None


class CameraRead(BaseModel):
    """Serialize a camera entity for API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    host: str
    port: int
    username: str
    profile_token: str | None
    snapshot_url: str | None
    interval_seconds: int
    enabled: bool
    created_at: datetime
    updated_at: datetime


class CameraTestResult(BaseModel):
    """Serialize the outcome of a camera connectivity test."""

    reachable: bool
    profiles: list[str] = []
    error: str | None = None


class SnapshotRead(BaseModel):
    """Serialize a snapshot entity for API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    camera_id: int
    captured_at: datetime
    image_path: str
    file_size: int
    status: str
    error_message: str | None


class SnapshotForceResult(BaseModel):
    """Serialize the result of a forced snapshot capture operation."""

    camera_id: int
    camera_name: str
    success: bool
    image_path: str | None = None
    error: str | None = None


class SnapshotAnalysisRead(BaseModel):
    """Serialize a snapshot analysis result for API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    snapshot_id: int
    model_name: str
    model_version: str
    status: str
    objects_json: str | None = None
    person_count: int = 0
    review_required: bool = False
    review_reason: str | None = None
    anomaly_score: float | None = None
    error_message: str | None = None
    analyzed_at: datetime | None = None


class SnapshotWithAnalysis(SnapshotRead):
    """Extend SnapshotRead with optional analysis data."""

    analysis: SnapshotAnalysisRead | None = None


class DailyReportCamera(BaseModel):
    """Serialize one camera's section within a daily report."""

    camera_id: int
    camera_name: str
    total_snapshots: int
    snapshots: list[SnapshotWithAnalysis]


class DailyReport(BaseModel):
    """Serialize the full daily snapshot report across all cameras."""

    date: str
    cameras: list[DailyReportCamera]


class CameraWithLastSnapshot(CameraRead):
    """Serialize a camera together with its most recent snapshot."""

    last_snapshot: SnapshotRead | None = None
    total_snapshots: int = 0


class VideoRequest(BaseModel):
    """Validate payload for timelapse video generation."""

    camera_id: int = Field(ge=1)
    date: date
    hour: int | None = Field(default=None, ge=0, le=23)


class VideoResponse(BaseModel):
    """Serialize the result of a successful video generation."""

    video_url: str


class AnalysisReviewUpdate(BaseModel):
    """Payload for updating the review status of an analysis."""

    review_required: bool
    review_reason: str | None = None


class RetentionResultRead(BaseModel):
    """Serialize the outcome of a retention/archive cleanup run.

    All counts are non-negative integers describing how many items
    were processed in each step of the retention lifecycle.
    """

    snapshots_zipped: int = Field(default=0, ge=0)
    snapshots_deleted: int = Field(default=0, ge=0)
    videos_archived: int = Field(default=0, ge=0)
    videos_deleted: int = Field(default=0, ge=0)


class PendingReviewItem(BaseModel):
    """A snapshot flagged for human review with its metadata."""

    model_config = ConfigDict(from_attributes=True)

    analysis_id: int
    snapshot_id: int
    camera_id: int
    camera_name: str
    captured_at: datetime
    image_path: str
    model_name: str
    person_count: int = 0
    review_required: bool = False
    review_reason: str | None = None
    anomaly_score: float | None = None
    error_message: str | None = None
    objects_json: str | None = None
    analyzed_at: datetime | None = None

File: internal/domain/schemas.go
package domain

// API request/response schemas mirroring the Pydantic v2 schemas of the
// legacy application. Validation happens in the handler layer (bounded
// fields, required strings) before reaching the services.

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// CameraCreate is the payload for creating a camera.
type CameraCreate struct {
	Name            string  `json:"name"`
	CameraType      string  `json:"camera_type"`
	Host            string  `json:"host"`
	Port            int     `json:"port"`
	Username        string  `json:"username"`
	Password        string  `json:"password"`
	ProfileToken    *string `json:"profile_token"`
	SnapshotURL     *string `json:"snapshot_url"`
	DevicePath      *string `json:"device_path"`
	IntervalSeconds int     `json:"interval_seconds"`
	Enabled         bool    `json:"enabled"`
}

// CameraUpdate is the payload for partially updating a camera. Pointer
// fields distinguish "unset" from "explicitly cleared".
type CameraUpdate struct {
	Name            *string `json:"name"`
	CameraType      *string `json:"camera_type"`
	Host            *string `json:"host"`
	Port            *int    `json:"port"`
	Username        *string `json:"username"`
	Password        *string `json:"password"`
	ProfileToken    *string `json:"profile_token"`
	SnapshotURL     *string `json:"snapshot_url"`
	DevicePath      *string `json:"device_path"`
	IntervalSeconds *int    `json:"interval_seconds"`
	Enabled         *bool   `json:"enabled"`
}

// CameraTestResult reports ONVIF reachability for an untested payload.
type CameraTestResult struct {
	Reachable bool     `json:"reachable"`
	Profiles  []string `json:"profiles"`
	Error     *string  `json:"error"`
}

// SnapshotWithAnalysis enriches a snapshot with its ML analysis.
type SnapshotWithAnalysis struct {
	Snapshot
	Analysis *SnapshotAnalysis `json:"analysis"`
}

// CameraWithLastSnapshot augments a camera with dashboard metrics.
type CameraWithLastSnapshot struct {
	Camera
	LastSnapshot   *Snapshot `json:"last_snapshot"`
	TotalSnapshots int       `json:"total_snapshots"`
}

// SnapshotForceResult reports an on-demand capture outcome.
type SnapshotForceResult struct {
	CameraID   int64   `json:"camera_id"`
	CameraName string  `json:"camera_name"`
	Success    bool    `json:"success"`
	ImagePath  *string `json:"image_path"`
	Error      *string `json:"error"`
}

// DailyReportCamera groups one camera's snapshots for a date.
type DailyReportCamera struct {
	CameraID       int64                  `json:"camera_id"`
	CameraName     string                 `json:"camera_name"`
	TotalSnapshots int                    `json:"total_snapshots"`
	Snapshots      []SnapshotWithAnalysis `json:"snapshots"`
}

// DailyReport is the full daily report response.
type DailyReport struct {
	Date    string              `json:"date"`
	Cameras []DailyReportCamera `json:"cameras"`
}

// Day is a calendar date serialized as YYYY-MM-DD (the legacy Pydantic
// date type). It accepts both "YYYY-MM-DD" and RFC3339 timestamps on
// input and always emits "YYYY-MM-DD" on output.
type Day time.Time

// UnmarshalJSON accepts "2006-01-02" (and RFC3339) values.
func (d *Day) UnmarshalJSON(data []byte) error {
	raw := strings.Trim(string(data), `"`)
	if raw == "" {
		return nil
	}
	for _, layout := range []string{"2006-01-02", time.RFC3339} {
		if t, err := time.ParseInLocation(layout, raw, time.Local); err == nil {
			*d = Day(t)
			return nil
		}
	}
	return fmt.Errorf("invalid date %q (want YYYY-MM-DD)", raw)
}

// MarshalJSON emits "YYYY-MM-DD".
func (d Day) MarshalJSON() ([]byte, error) {
	return json.Marshal(time.Time(d).Format("2006-01-02"))
}

// Time converts the day to a time.Time at midnight in the project zone.
func (d Day) Time() time.Time { return time.Time(d) }

// VideoRequest asks for a plain timelapse for a camera/date.
type VideoRequest struct {
	CameraID int64 `json:"camera_id"`
	Date     Day   `json:"date"`
	Hour     *int  `json:"hour"`
}

// VideoResponse returns the download URL of a generated video.
type VideoResponse struct {
	VideoURL string `json:"video_url"`
}

// AnnotatedVideoRequest asks for an annotated timelapse.
type AnnotatedVideoRequest struct {
	CameraID int64   `json:"camera_id"`
	Date     Day     `json:"date"`
	Classes  *string `json:"classes"`
}

// AnnotatedVideoResponse mirrors VideoResponse for the annotated flow.
type AnnotatedVideoResponse struct {
	VideoURL string `json:"video_url"`
}

// AnalysisReviewUpdate changes the review flag of an analysis.
type AnalysisReviewUpdate struct {
	ReviewRequired bool    `json:"review_required"`
	ReviewReason   *string `json:"review_reason"`
}

// BulkReviewPayload updates several analyses at once.
type BulkReviewPayload struct {
	AnalysisIDs    []int64 `json:"analysis_ids"`
	ReviewRequired bool    `json:"review_required"`
	ReviewReason   *string `json:"review_reason"`
}

// PendingReviewItem lists one flagged snapshot for human review.
type PendingReviewItem struct {
	AnalysisID     int64    `json:"analysis_id"`
	SnapshotID     int64    `json:"snapshot_id"`
	CameraID       int64    `json:"camera_id"`
	CameraName     string   `json:"camera_name"`
	CapturedAt     SQLTime  `json:"captured_at"`
	ImagePath      string   `json:"image_path"`
	ModelName      string   `json:"model_name"`
	PersonCount    int      `json:"person_count"`
	ReviewRequired bool     `json:"review_required"`
	ReviewReason   *string  `json:"review_reason"`
	AnomalyScore   *float64 `json:"anomaly_score"`
	ErrorMessage   *string  `json:"error_message"`
	ObjectsJSON    *string  `json:"objects_json"`
	AnalyzedAt     SQLTime  `json:"analyzed_at"`
}

// PurgeRequest asks to delete everything older than N days.
type PurgeRequest struct {
	Days int `json:"days"`
}

// Detection is one YOLO detection result.
type Detection struct {
	ClassName  string    `json:"class_name"`
	Confidence float64   `json:"confidence"`
	BBox       []float64 `json:"bbox"`
}
